text/3373-avoid-nonlocal-definitions-in-fns.md

```diff
@@ -0,0 +1,71 @@
+- Feature Name: N/A
+- Start Date: 2022-01-19
+- RFC PR: [rust-lang/rfcs#3373](https://github.com/rust-lang/rfcs/pull/3373)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Starting in Rust 2024, stop allowing items inside functions or expressions to
+implement methods or traits that are visible outside the function or
+expression.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, tools cross-referencing uses and definitions (such as IDEs) must
+either search inside all function bodies and other expression-containing items
+to find potential definitions corresponding to uses within another function, or
+not cross-reference those definitions at all.
+
+Humans cross-referencing such uses and definitions may find themselves
+similarly baffled.
+
+With this change, both humans and tools can limit the scope of their search and
+avoid looking for definitions inside other functions or items, without missing
+any relevant definitions.
+
+# Explanation
+[explanation]: #explanation
+
+The following types of items, "expression-containing items", can contain
+expressions, including the definitions of other items:
+- Functions
+- Closures
+- The values assigned to `static`/`const` items
```

Comment:
Forbidding non-local definitions in all kinds of `const`s sounds promising, but IIRC many `derive` macros uses `const _: () = { ... }` anonymous constants for scoping.

Reply:
If we stabilize inline `const {}` this is not as much of a concern, I think?

Reply:
no, because inline const is an expression, not an item, so the derive macro can't use it because the derive could be invoked at global scope where inline const is invalid, because only items are permitted there.